doc/man/gnome-session-inhibit.1
'\" t
.\"     Title: gnome-session-inhibit
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.77.1 <http://docbook.sf.net/>
.\"      Date: 11/07/2012
.\"    Manual: User Commands
.\"    Source: gnome-session
.\"  Language: English
.\"
.TH "GNOME\-SESSION\-INHI" "1" "11/07/2012" "gnome-session" "User Commands"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gnome-session-inhibit \- inhibit gnome\-session functionality
.SH "SYNOPSIS"
.HP \w'\fBgnome\-session\-inhibit\fR\ 'u
\fBgnome\-session\-inhibit\fR [OPTION...] [COMMAND]
.SH "DESCRIPTION"
.PP
\fBgnome\-session\-inhibit\fR
can inhibit certain gnome\-session functionality while executing the given COMMAND\&. To achieve this, it calls the Inhibit() method of the gnome\-session D\-Bus API and creates an inhibitor\&. The inhibitor is automatically removed when gnome\-session\-inhibit exits\&.
.PP
A typical use case is to prevent the session from going idle (and thus locking the screen) while a movie player is running\&.
.SH "OPTIONS"
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
print help and exit
.RE
.PP
\fB\-\-version\fR
.RS 4
print version information and exit
.RE
.PP
\fB\-\-app\-id\fR ID
.RS 4
The application id to use when calling the gnome\-session Inhibit() method\&. If this option is not specified, "unknown" is used\&.
.RE
.PP
\fB\-\-reason\fR REASON
.RS 4
A human\-readable reason to pass along when calling the gnome\-session Inhibit() method\&. If this option is not specified, "not specified" is used\&.
.RE
.PP
\fB\-\-inhibit\fR ARG
.RS 4
ARG specifies the things to inhibit, as a colon\-separated list\&. The possible values are logout, switch\-user, suspend, idle, automount\&. If this option is used more than once, the values are combined\&. If this option is not specified, "idle" is assumed\&.
.RE
.PP
\fB\-\-inhibit\-only\fR
.RS 4
Do not launch COMMAND and wait forever instead
.RE
.SH "SEE ALSO"
.PP

\fBsystemd-inhibit\fR(1)


NEWS
=============
Version 3.7.1
=============

 Session Manager

	* Add gnome-screensaver to fallback session
	* Make SessionName property introspectable

=============
Version 3.6.0
=============

 Session Manager

	* Make autostart overriding more flexible
	* Updated translations

==============
Version 3.5.91
==============

 Session Manager

	* Updated translations

=============
Version 3.5.5
=============

 Session Manager

	* Make XSMP error messages more clear
	* Be more careful in systemd code when session fails to get registered
	* Require newer glib
	* Make gconf build time optional
	* Updated translations

=============
Version 3.5.4
=============

 Session Manager

	* Add command for showing reboot dialog
	* Inhibitor API / mechanism improvements
        * Systemd integration fixes
	* man page updates
	* fail whale in more failure cases
	* fix duplicate logout dialog when user hits cancel
	* Updated translations

=============
Version 3.5.2
=============

 Session Manager

	* Require new logind version (Colin Walters)
	* Remove markup from translatable string (Piotr Drąg)
	* Fail softer in the face of spotty session bus (Ray, William Jon McCann)
	* Consider that required component that exits with 1 has crashed (Vincent)

  Translators

	* Dirgita (id)

=============
Version 3.4.2
=============

  Capplet

	* Add Unity to OnlyShowIn in .desktop file (Jeremy Bicha)

  Translators

	* Yurek Hinz (csb)
	* Takayuki KUSANO (ja)

=============
Version 3.4.1
=============

  Session Manager

	* Properly move to next phase if an app dies on startup (Vincent)
	* Fix some restart/poweroff problems in the systemd code (Matthias
	  Clasen)
	* Cleanups (Vincent)

  Translators

	* David Planella (ca)
	* David Planella (ca@valencia)
	* Reşat SABIQ (crh)
	* Wolfgang Stoeggl (de)
	* Bruce Cowan (en_GB)
	* Arash Mousavi (fa)
	* Gabor Kelemen (hu)
	* Milo Casagrande (it)
	* Sandeep Shedmake (mr)
	* Kjartan Maraas (nb)

=============
Version 3.4.0
=============

  Session Manager

	* Don't show toggle switches for gnome-shell extensions in fail whale
	  dialog; simply disable all extensions instead (Jasper St. Pierre)
	* Code cleanups (Jasper St. Pierre)

  Translators

	* Nilamdyuti Goswami (as)
	* Ihar Hrachyshka (be)
	* Alexander Shopov (bg)
	* David Planella (ca)
	* David Planella (ca@valencia)
	* Reşat SABIQ (crh)
	* Marek Černocký (cs)
	* flemming christensen (da)
	* Daniel Mustieles (es)
	* Mattias Põldaru (et)
	* Luc Guillemin (fr)
	* Fran Dieguez (gl)
	* Sweta Kothari (gu)
	* Yaron Shahrabani (he)
	* Changwoo Ryu (ko)
	* Aurimas Černius (lt)
	* Rūdolfs Mazurs (lv)
	* A S Alam (pa)
	* Piotr Drąg (pl)
	* Duarte Loreto (pt)
	* Antonio Fernandes C. Neto (pt_BR)
	* Yuri Myasoedov (ru)
	* Matej Urbančič (sl)
	* Мирослав Николић (sr)
	* Miroslav Nikolić (sr@latin)
	* Daniel Nylander (sv)
	* Sasi Bhushan Boddepalli (te)
	* Korostil Daniel (uk)
	* Nguyễn Thái Ngọc Duy (vi)
	* tuhaihe (zh_CN)
	* Chao-Hsiung Liao (zh_HK)
	* Chao-Hsiung Liao (zh_TW)

==============
Version 3.3.92
==============

  Session Manager

	* Minor bug fixes (Ray Strode, Jasper St. Pierre)

  Translators

	* Nilamdyuti Goswami (as)
	* Mattias Põldaru (et)
	* Yuri Myasoedov (ru)

==============
Version 3.3.90
==============

  Session Manager

	* Fix the GsmSystem type registration (Matthias Clasen)

=============
Version 3.3.5
=============

  Session Manager

	* Use systemd session tracking instead of ConsoleKit when system is
	  booted with systemd (Matthias Clasen)
	* Code cleanups (Matthias Clasen)

  Translators

	* Hideki Yamane (ja)
	* Andreas N. (nn)
	* Muhammet Kara (tr)

=============
Version 3.3.3
=============

  Tools

	* Blacklist pre-R300 Radeon hardware in accelerated check (Vincent)

=============
Version 3.3.2
=============

  Tools

	* Fix regexp for blacklisting Intel 830-865 hardware (Vincent)

  Translators

	* Kristjan SCHMIDT (eo)
	* Muhammet Kara (tr)

=============
Version 3.3.1
=============

  Tools

	* Look at gnome.fallback={0,1} argument in kernel boot line to
	  determine if fallback mode should be enforced or ignored (Vincent)
	* Use whitelist/blacklist regexps in an external file when evaluating
	  renderer for accelerated check (Vincent)
	* Blacklist Intel 830-865 hardware when checking for accelerated
	  hardware (Vincent)

  Translators

	* Ihar Hrachyshka (be)

=============
Version 3.2.1
=============

  Session Manager

	* Refuse to install an idle timeout of 0 seconds (Rodrigo Moya)
	* Fix crash on Solaris because of bad comparison (Brian Cameron)

  Translators

	* David Planella (ca@valencia)
	* Kristjan SCHMIDT (eo)
	* Nguyễn Thái Ngọc Duy (vi)
	* Chao-Hsiung Liao (zh_HK)
	* Chao-Hsiung Liao (zh_TW)

=============
Version 3.2.0
=============

  Translators

	* Nilamdyuti Goswami (as)
	* Petr Kovar (cs)
	* flemming christensen (da)
	* Inaki Larranaga Murgoitio (eu)
	* Tommi Vainikainen (fi)
	* Sweta Kothari (gu)
	* Gabor Kelemen (hu)
	* Andika Triwidada (id)
	* Changwoo Ryu (ko)
	* Aurimas Černius (lt)
	* Мирослав Николић (sr)
	* Miroslav Nikolić (sr@latin)
	* Praveen Illa (te)
	* Korostil Daniel (uk)

==============
Version 3.1.92
==============

  Session Manager

	* Never mark as handled the Disconnected signal from dbus (Vincent)
	* Disconnect all dbus clients when dbus is disconnected (Vincent)
	* Don't offer logout in the fail whale on the login screen (Matthias Clasen)
	* Code improvements (Matthias Clasen, Vincent)

  Translators

	* Anass Ahmed (ar)
	* Ihar Hrachyshka (be)
	* Alexander Shopov (bg)
	* David Planella (ca)
	* Mario Blättermann (de)
	* Bruce Cowan (en_GB)
	* Daniel Mustieles (es)
	* Arash Mousavi (fa)
	* Luc Guillemin (fr)
	* Yaron Shahrabani (he)
	* Luca Ferretti (it)
	* Jiro Matsuzawa (ja)
	* Rūdofls Mazurs (lv)
	* Wouter Bolsterlee (nl)
	* A S Alam (pa)
	* Piotr Drąg (pl)
	* Duarte Loreto (pt)
	* Antonio Fernandes C. Neto (pt_BR)
	* Matej Urbančič (sl)
	* Dr.T.Vasudevan (ta)
	* Theppitak Karoonboonyanan (th)
	* Wylmer Wang (zh_CN)
	* Chao-Hsiung Liao (zh_HK)

tools/gnome-session-inhibit.c
/*
 * Copyright (C) 2012 Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */

#include "config.h"

#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <locale.h>

#include <glib.h>
#include <glib/gi18n.h>
#include <gio/gio.h>

#include "gnome-session/gsm-inhibitor-flag.h"

static GsmInhibitorFlag parse_flags (const gchar *arg)
{
  GsmInhibitorFlag flags;
  gchar **args;
  gint i;

  flags = 0;

  args = g_strsplit (arg, ":", 0);
  for (i = 0; args[i]; i++)
    {
      if (strcmp (args[i], "logout") == 0)
        flags |= GSM_INHIBITOR_FLAG_LOGOUT;
      else if (strcmp (args[i], "switch-user") == 0)
        flags |= GSM_INHIBITOR_FLAG_SWITCH_USER;
      else if (strcmp (args[i], "suspend") == 0)
        flags |= GSM_INHIBITOR_FLAG_SUSPEND;
      else if (strcmp (args[i], "idle") == 0)
        flags |= GSM_INHIBITOR_FLAG_IDLE;
      else if (strcmp (args[i], "automount") == 0)
        flags |= GSM_INHIBITOR_FLAG_AUTOMOUNT;
      else
        g_print ("Ignoring inhibit argument: %s\n", args[i]);
    }

  g_strfreev (args);

  return flags;
}

static gboolean inhibit (const gchar       *app_id,
                         const gchar       *reason,
                         GsmInhibitorFlag flags)
{
  GDBusConnection *bus;
  GVariant *ret;
  GError *error = NULL;

  bus = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, &error);

  if (bus == NULL)
    {
      g_warning ("Failed to connect to session bus: %s", error->message);
      g_error_free (error);
      return FALSE;
    }

  ret = g_dbus_connection_call_sync (bus,
                               "org.gnome.SessionManager",
                               "/org/gnome/SessionManager",
                               "org.gnome.SessionManager",
                               "Inhibit",
                               g_variant_new ("(susu)",
                                              app_id, 0, reason, flags),
                               G_VARIANT_TYPE ("(u)"),
                               0,
                               G_MAXINT,
                               NULL,
                               &error);

  if (ret == NULL)
    {
      g_warning ("Failed to call Inhibit: %s\n", error->message);
      g_error_free (error);
      return FALSE;
    }

  g_variant_unref (ret);

  return TRUE;
}

static void usage (void)
{
  g_print (_("%s [OPTION...] COMMAND\n"
             "\n"
             "Execute COMMAND while inhibiting some session functionality.\n"
             "\n"
             "  -h, --help        Show this help\n"
             "  --version         Show program version\n"
             "  --app-id ID       The application id to use\n"
             "                    when inhibiting (optional)\n"
             "  --reason REASON   The reason for inhibiting (optional)\n"
             "  --inhibit ARG     Things to inhibit, colon-separated list of:\n"
             "                    logout, switch-user, suspend, idle, automount\n"
             "  --inhibit-only    Do not launch COMMAND and wait forever instead\n"
             "\n"
             "If no --inhibit option is specified, idle is assumed.\n"),
           g_get_prgname ());
}

static void version (void)
{
  g_print ("%s %s\n", g_get_prgname (), PACKAGE_VERSION);
}

static void
wait_for_child_app (char **argv)
{
  pid_t pid;
  int status;

  pid = fork ();
  if (pid < 0)
    {
      g_print ("fork failed\n");
      exit (1);
    }

  if (pid == 0)
    {
      execvp (*argv, argv + 1);
      g_print (_("Failed to execute %s\n"), *argv);
      exit (1);
    }

  do
    {
      if (waitpid (pid, &status, 0) == -1)
        {
          g_print ("waitpid failed\n");
          exit (1);
        }

      if (WIFEXITED (status))
        exit (WEXITSTATUS (status));

    } while (!WIFEXITED (status) && ! WIFSIGNALED(status));
}

static void
wait_forever (void)
{
  g_print ("Inhibiting until Ctrl+C is pressed...\n");
  while (1)
    sleep (UINT32_MAX);
}

int main (int argc, char *argv[])
{
  gchar *prgname;
  GsmInhibitorFlag inhibit_flags = 0;
  gboolean show_help = FALSE;
  gboolean show_version = FALSE;
  gboolean no_launch = FALSE;
  gint i;
  const gchar *app_id = "unknown";
  const gchar *reason = "not specified";

  prgname = g_path_get_basename (argv[0]);
  g_set_prgname (prgname);
  g_free (prgname);

  setlocale (LC_ALL, "");
  bindtextdomain (GETTEXT_PACKAGE, LOCALE_DIR);
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
  textdomain (GETTEXT_PACKAGE);

  for (i = 1; i < argc; i++)
    {
      if (strcmp (argv[i], "--help") == 0 ||
          strcmp (argv[i], "-h") == 0)
        show_help = TRUE;
      else if (strcmp (argv[i], "--version") == 0)
        show_version = TRUE;
      else if (strcmp (argv[i], "--inhibit-only") == 0)
        no_launch = TRUE;
      else if (strcmp (argv[i], "--app-id") == 0)
        {
          i++;
          if (i == argc)
            {
              g_print (_("%s requires an argument\n"), argv[i]);
              exit (1);
            }
          app_id = argv[i];
        }
      else if (strcmp (argv[i], "--reason") == 0)
        {
          i++;
          if (i == argc)
            {
              g_print (_("%s requires an argument\n"), argv[i]);
              exit (1);
            }
          reason = argv[i];
        }
      else if (strcmp (argv[i], "--inhibit") == 0)
        {
          i++;
          if (i == argc)
            {
              g_print (_("%s requires an argument\n"), argv[i]);
              exit (1);
            }
          inhibit_flags |= parse_flags (argv[i]);
        }
      else
        break;
    }

  if (show_version)
    {
      version ();
      return 0;
    }

  if (show_help || (i == argc && !no_launch))
    {
      usage ();
      return 0;
    }

  if (inhibit_flags == 0)
    inhibit_flags = GSM_INHIBITOR_FLAG_IDLE;

  if (inhibit (app_id, reason, inhibit_flags) == FALSE)
    return 1;

  if (!no_launch)
    wait_for_child_app (argv + i);
  else
    wait_forever ();

  return 0;
}
